### PR TITLE
Resolves domain base types

### LINKED ISSUE #399: Better domain type support
**Describe the bug**

Types for scalar domain types return type Opaque and for array domain types they just fail.

**To Reproduce**

Given the following schema:

```sql
create domain pos_int as int check (value > 0);
create domain non_empty_array as int[] check (array_length(value, 1) > 0);
create domain pos_array as pos_int[];

create table domain_test(
    id serial primary key,
    int_field int,
    array_field int[],
    pos_int pos_int,
    non_empty_array non_empty_array,
    pos_array pos_array
);

insert into domain_test(int_field, array_field, pos_int, non_empty_array, pos_array)
values (1, '{1,2,3}', 1, '{1,2,3}', '{1,2,3}');
```

Querying for the types on `domain_test` returns the following:

```sql
select graphql.resolve($$
query {
  __type(name: "domain_test") {
    fields {
      name
      type {
        name
        kind
        ofType {
          name
          kind
          ofType {
            name
            kind
          }
        }
      }
    }
  }
}
$$);
```

```json
{
  "data": {
    "__type": {
      "fields": [
        {
          "name": "nodeId",
          "type": {
            "kind": "NON_NULL",
            "name": null,
            "ofType": {
              "kind": "SCALAR",
              "name": "ID",
              "ofType": null
            }
          }
        },
        {
          "name": "id",
          "type": {
            "kind": "NON_NULL",
            "name": null,
            "ofType": {
              "kind": "SCALAR",
              "name": "Int",
              "ofType": null
            }
          }
        },
        {
          "name": "int_field",
          "type": {
            "kind": "SCALAR",
            "name": "Int",
            "ofType": null
          }
        },
        {
          "name": "array_field",
          "type": {
            "kind": "LIST",
            "name": null,
            "ofType": {
              "kind": "SCALAR",
              "name": "Int",
              "ofType": null
            }
          }
        },
        {
          "name": "pos_int",
          "type": {
            "kind": "SCALAR",
            "name": "Opaque",
            "ofType": null
          }
        }
      ]
    }
  }
}
```

Querying the data will work for the fields on the Opaque types:

```sql
select graphql.resolve($$
query {
  domain_testCollection{
    edges {
        node {
            id
            int_field
            array_field
            pos_int
        }
    }
  }
}
$$);
```

```json
{
  "data": {
    "domain_testCollection": {
      "edges": [
        {
          "node": {
            "id": 1,
            "pos_int": 1,
            "int_field": 1,
            "array_field": [
              1,
              2,
              3
            ]
          }
        }
      ]
    }
  }
}
```

And fails for fields with an array domain_type:

```sql
select graphql.resolve($$
query {
  domain_testCollection{
    edges {
        node {
            id
            non_empty_array
        }
    }
  }
}
$$);
```
```json
{
  "data": null,
  "errors": [
    {
      "message": "Unknown field 'non_empty_array' on type 'domain_test'"
    }
  ]
}
```

You can attempt the above with `pos_array` and get a similar result.

**Expected behavior**

For fields with domain types to resolve to there domain type or at least the domain base type.

E.g. 

```sql
select graphql.resolve($$
query {
  __type(name: "domain_test") {
    fields {
      name
      type {
        name
        kind
        ofType {
          name
          kind
          ofType {
            name
            kind
          }
        }
      }
    }
  }
}
$$);
```

Would return:

```json
{
  "data": {
    "__type": {
      "fields": [
        {
          "name": "nodeId",
          "type": {
            "kind": "NON_NULL",
            "name": null,
            "ofType": {
              "kind": "SCALAR",
              "name": "ID",
              "ofType": null
            }
          }
        },
        {
          "name": "id",
          "type": {
            "kind": "NON_NULL",
            "name": null,
            "ofType": {
              "kind": "SCALAR",
              "name": "Int",
              "ofType": null
            }
          }
        },
        {
          "name": "int_field",
          "type": {
            "kind": "SCALAR",
            "name": "Int",
            "ofType": null
          }
        },
        {
          "name": "array_field",
          "type": {
            "kind": "LIST",
            "name": null,
            "ofType": {
              "kind": "SCALAR",
              "name": "Int",
              "ofType": null
            }
          }
        },
        {
          "name": "pos_int",
          "type": {
            "kind": "SCALAR",
            "name": "Int",
            "ofType": null
          }
        },
        {
          "name": "non_empty_array",
          "type": {
            "kind": "LIST",
            "name": null,
            "ofType": {
              "kind": "SCALAR",
              "name": "Int",
              "ofType": null
            }
          }
        },
        {
          "name": "pos_array",
          "type": {
            "kind": "LIST",
            "name": null,
            "ofType": {
              "kind": "SCALAR",
              "name": "Int",
              "ofType": null
            }
          }
        }
      ]
    }
  }
}
```

The behavior of the field itself on querying would be the same as it's base type.

**Screenshots**
N/A

**Versions:**
 - PostgreSQL: 15
 - pg_graphql commit ref: ec3f7d5473747cc7a0c9e2495b1986adffdfef86

**Additional context**
I think I have a possible fix for this and I'l like to give opening a pr to fix this.

A simple solution would be to resolve the base type of the domain object itself.
To my understanding domain types always have to eventually resolve to a built in sql scalar or array type so this seems like a valid solution.
While this would technically be a breaking change Opaque isn't particularly useful so it could also hide behind a schema directive flag but that would be opt out. This is also a possible fix for #370.

Let me know if you agree and I'll submit a pr.

The complete  fix would be to generate new types for domain types.
That is breaking change and some client might not care to know about anything but the base type. It'll have to be behind a schema directive flag that's opt in.
I won't be attempting this approach right away just spitballing a final solution.

### PR DESCRIPTION
**Pull request by andrew-w-ross**
Closes #399 ~~and potentially #370~~.

_Looks like `citext` type doesn't extend text this pr won't fix it._ 

I think the coding portion of this is done.
Still need to finish off the domain type tests and the domain array type tests.

Slight deviation is that array types resolve as `[base_type]` instead of `[Opaque]`.

Edit: I knew that felt too easy, looks like mutation updates have some bugs but for some odd reason mutation insert works fine. 

Edit: That was a lot harder then expected a lot of edge cases to deal with. 

I'm going to mark this as ready for review but I think there might be potential tests for nullability in domain types.

E.g. 

```sql
create domain not_null_int as int not null;

create table not_null_test (
    id serial primary key,
    field_int not_null_int
 );
```

Will resolve with a `graphql` type of:

```graphql
type NotNullTest {
 id: Int!
 fieldInt: Int
}
``` 

Considering this is a bit of an anti pattern, check the [notes section for create domain](https://www.postgresql.org/docs/current/sql-createdomain.html), it might not be worth dealing with.

### CODE REVIEW

File: sql/load_sql_context.sql [hunk 1]
```diff
@@ -192,6 +221,10 @@ select
                                 'max_rows', coalesce(
                                     (graphql.comment_directive(pg_catalog.obj_description(pn.oid, 'pg_namespace')) ->> 'max_rows')::int,
                                     30
+                                ),
+                                'resolve_base_type', coalesce(
+                                    (graphql.comment_directive(pg_catalog.obj_description(pn.oid, 'pg_namespace')) -> 'resolve_domain_types') = to_jsonb(true),
+                                    true
```
Comment (imor) [inline sql/load_sql_context.sql:227]: Shouldn't the default be `false`?

File: test/expected/domain_field_types.out [hunk 2]
```diff
@@ -0,0 +1,547 @@
...
+             "edges": [                                          +
+                 {                                               +
+                     "node": {                                   +
+                         "id": 1,                                +
+                         "fieldInt": 42,                         +
+                         "fieldDate": "2022-01-01",              +
+                         "fieldReal": 3.14,                      +
+                         "fieldText": "Hello, world!",           +
+                         "fieldTime": "12:34:56",                +
+                         "fieldBigint": 2165559898978,           +
```
Comment (imor) [inline test/expected/domain_field_types.out:211]: BigInt is being serialized as json int. It should be serialized as a string.

File: test/sql/domain_field_types.sql [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+        }
+      }
+    $$)
+  );
+
+  rollback to savepoint a;
+
+  -- TODO: Check that update mutation resolves the base types
+  -- TODO: Check that delete mutation resolves the base types
+
```
Comment (imor) [inline test/sql/domain_field_types.sql:197]: Current tests are written for when `resolve_base_type` is true. It's worth adding tests for when `resolve_base_type` is false and for a [computed column that uses a function](https://supabase.github.io/pg_graphql/computed_fields/) and returns a domain type.

File: test/sql/domain_field_types.sql [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+rollback;
```
Comment (imor) [inline test/sql/domain_field_types.sql:198]: Docs also need to be updated with the new directive.

File: test/sql/domain_field_types.sql [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+      field_real domain_real,
+      field_double_precision domain_double_precision,
+      field_boolean domain_boolean,
+      field_text domain_text,
+      field_character domain_character,
+      field_character_varying domain_character_varying,
+      field_date domain_date,
+      field_time domain_time,
+      field_timestamp domain_timestamp
+  );
```
Comment (imor) [inline test/sql/domain_field_types.sql:32]: Worth adding domain types on `json` and `jsonb`. Both these should be serialized as strings in GraphQL.

File: test/expected/domain_field_types.out [hunk 6]
```diff
@@ -0,0 +1,547 @@
...
+                         "fieldBoolean": true,                   +
+                         "fieldNumeric": 123.45,                 +
```
Comment (imor) [inline test/expected/domain_field_types.out:213]: Same as BigInt.

File: sql/load_sql_context.sql (hunk 1 above)
Comment (andrew-w-ross) [reply sql/load_sql_context.sql:227]: I thought we were going with this minor change you have to opt out of but I have no issue making it opt into.

File: test/sql/domain_field_types.sql (hunk 5 above)
Comment (andrew-w-ross) [reply test/sql/domain_field_types.sql:32]: Good catch, didn't even think about it.

File: sql/load_sql_context.sql (hunk 1 above)
Comment (andrew-w-ross) [reply sql/load_sql_context.sql:227]: On second thought I think you're right, 2.0 it can change to the default.

**Comment by andrew-w-ross** [comment]
@imor Should I update this pr or is there no hope in getting it in?

**Comment by imor** [comment]
Hey @andrew-w-ross I thought this was still in progress. If this was ready, ideally I'd re-request review from the reviewers section at the top right of the PR. Anyways, I've just fixed a conflict and will review by tomorrow.

**Comment by andrew-w-ross** [comment]
@imor Thanks, will do next time

**Review by imor — COMMENTED** [review]
Thanks @andrew-w-ross for this PR. Looking good. Just one question about functions.

@olirice please take a quick look at this PR, if everything looks good we can merge.

File: src/sql_types.rs [hunk 7]
```diff
@@ -540,6 +543,79 @@ pub fn load_sql_context(_config: &Config) -> Result<Arc<Context>, String> {
...
+                };
+
+                if let Some(column) = Arc::get_mut(column) {
+                    column.type_oid = base_oid;
+                }
+            }
+
+            //Technically speaking, we should check the schema of the function to see if we should resolve the base type
+            //Problem is this might be counter-intuitive to the user as in this case the function is acting as a virtual column of the table
+            for function in table.functions.iter_mut() {
```
[the remark this answers is not in the document]
Comment (andrew-w-ross) [reply src/sql_types.rs:604]: The use case is for users that separate virtual columns into a private schema to try and reduce noise on the public schema. Here is the setup:

```sql
create schema private;

comment on schema public is e'@graphql({"inflect_names": true, "resolve_base_type": true})';
-- Not needed, this is the default but doing for illustration
comment on schema private is e'@graphql({"inflect_names": true, "resolve_base_type": false})';

create table notes (
  id serial primary key,
  title text not null,
  body text not null,
  created_at timestamp not null default now()
);

create domain natural_int as int check (value >= 0); 

create or replace function private._age_seconds(notes) returns natural_int as $$
  select extract(epoch from (now() - $1.created_at ))::natural_int;
$$ language sql strict;

select 
  graphql.resolve (
    $$
      {
        __type(name: "Notes") {
          kind
          fields {
              name
              type {
                  name
                  kind
              }
          }
        }
      }
    $$
  );
```

If you query the schema types for the model as it currently stands  the base type resolves for all fields including the virtual. I suspect most users mental model consider the virtual columns  to be an extension of the table so following the same rules the table does making this intuitive. 

If instead I checked for the function schema settings then all your fields would resolve to the base type and the virtual fields would resolve to opaque. I think most users would be confused or just chalk this up to a bug. 

I am assuming that most users consider virtual columns to be an extension to the table they're referencing but it's the mental model I've observed other devs gravitating to.

[the remark this answers is not in the document]
Comment (andrew-w-ross) [reply src/sql_types.rs:604]: As for the documentation I did make a note [of it here](https://github.com/supabase/pg_graphql/pull/402/files#diff-17ed18489a956f326ec0fe4040850c5bc9261d4631fb42da4c52891d74a59180R108-R111) as it follows the same behaviour of where the table is defined is the setting that matters not the domain. But if you'd like a specific callout to virtual columns let me know and I'll add it to the docs.

File: sql/load_sql_context.sql [hunk 8]
```diff
@@ -15,6 +15,17 @@ schemas_(oid, name) as (
...
+type_hierarchy(base_oid, current_oid) AS (
+    SELECT oid, oid
+    FROM pg_type
+    WHERE typbasetype = 0
+
+    UNION ALL
+
+    SELECT th.base_oid, pt.oid
+    FROM pg_type pt
+    JOIN type_hierarchy th ON pt.typbasetype = th.current_oid
```
Comment (olirice) [inline sql/load_sql_context.sql:28]: nit: match existing SQL formatting style (lowercase w/ newlines)

File: docs/configuration.md [hunk 9]
```diff
@@ -63,10 +63,96 @@ For more fine grained adjustments to reflected names, see [renaming](#renaming).
...
+The `age` field is an `Opaque` type:
+
+```graphql
+type Users {
+  id: ID!
+  age: Opaque
+}
+```
+
+Set the `resolve_base_type` directive to `true` to resolve the domain types to their base types instead. For example:
```
Comment (olirice) [inline docs/configuration.md:93]: How about `resolve_domains_as_base_type` for clarity?

File: test/expected/domain_field_types.out [hunk 10]
```diff
@@ -0,0 +1,1092 @@
...
+                     "fieldDoublePrecision": 3.14159265359,  +
+                     "fieldCharacterVarying": "ABCD"         +
+                 }                                           +
+             ],                                              +
+             "affectedCount": 1                              +
+         }                                                   +
+     }                                                       +
+ }
+(1 row)
+
```
Comment (olirice) [inline test/expected/domain_field_types.out:1091]: Could you add a tests that show what kind of error output we get if a user provides an invalid input value for a domain field?

For example, given a `pos_int` domain:
- what happens if the user enters a negative number in a Filter
- what happens if the user tries to insert a negative value

**Review by olirice — CHANGES_REQUESTED** [review]
great stuff! thanks for putting so much thought into the tests

a few nits but conceptually I'd love to get this merged

File: src/sql_types.rs [hunk 11]
```diff
@@ -540,6 +543,79 @@ pub fn load_sql_context(_config: &Config) -> Result<Arc<Context>, String> {
...
+
+                if let Some(function) = Arc::get_mut(function) {
+                    function.type_oid = base_oid;
+                }
+            }
+        }
+
+        context
+    }
+
```
[the remark this answers is not in the document]
Comment (andrew-w-ross) [reply src/sql_types.rs:618]: @olirice Fair point I was trying to change as little as possible initially but that's been thrown outside of the window at this point. It makes sense to include it as there is a decent amount more that can be done with that data. I'll get to it this weekend but if you don't mind just review this outline and correct me if I misunderstood something.

Change the [base_type_map to domain instead](https://github.com/andrew-w-ross/pg_graphql/blob/feature/domain-types/sql/load_sql_context.sql#L37C10-L54) instead and pull out all relevant, I'm thinking the oid, schema_oid, name, nullability, and possibly the element_oid. Might add the base_oid to that type to help with traversing the types later.

Change the types to include `Domain` in [category type](https://github.com/andrew-w-ross/pg_graphql/blob/feature/domain-types/sql/load_sql_context.sql#L105-L112) and adding it to the [TypeCategory](https://github.com/andrew-w-ross/pg_graphql/blob/feature/domain-types/src/sql_types.rs#L86C1-L93C2) enum. Also create a Domain struct and add it to [TypeDetails](https://github.com/andrew-w-ross/pg_graphql/blob/feature/domain-types/src/sql_types.rs#L111C1-L116C2).

Revert the [replace functions](https://github.com/andrew-w-ross/pg_graphql/blob/feature/domain-types/src/sql_types.rs#L546-L617) in `load_sql_context`. Mapping the domain type in the [type details function](https://github.com/andrew-w-ross/pg_graphql/blob/feature/domain-types/src/sql_types.rs#L629-L647). 

Finally adding another match block to for `Domain` in the [to_graphql_type](https://github.com/andrew-w-ross/pg_graphql/blob/feature/domain-types/src/graphql.rs#L1624-L1639) function.

As for the tests, you're happy with them as is so if they pass we're good right?

File: sql/load_sql_context.sql (hunk 8 above)
Comment (andrew-w-ross) [reply sql/load_sql_context.sql:28]: Sure, thing. 

Just thinking out loud  but it might be a good idea to add a formatter for the sql code. I haven't used pre-commit before so I don't know how difficult it'll be to add. But what about adding [pgFormatter](https://www.postgresql.org/about/news/pgformatter-54-released-2569/) until the supabase/postgres_lsp can handle formatting?

File: test/expected/domain_field_types.out (hunk 10 above)
Comment (andrew-w-ross) [reply test/expected/domain_field_types.out:1091]: Will do, Is this just for documenting behavior or is there a specific expected behavior other then returning:

```json
 { 
     "data": null,
     "errors": [ 
         {
             "message": "Something bad happened."
         }
     ]
 }
```

File: src/sql_types.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (imor) [reply src/sql_types.rs:604]: Ok, got it, thanks. I agree that it is better DX if I could just enable base type resolution on the table's schema and get both my real & virtual fields behave the same way. Could you please mention it in the docs that for virtual field resolution the directive should be on the table's schema not the function's.

File: src/sql_types.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (olirice) [reply src/sql_types.rs:618]: > Change the [base_type_map to domain instead](https://github.com/andrew-w-ross/pg_graphql/blob/feature/domain-types/sql/load_sql_context.sql#L37C10-L54) and pull out all relevant, I'm thinking the oid, schema_oid, name, nullability, and possibly the element_oid. Might add the base_oid to that type to help with traversing the types later.

sorry I didn't understand what you're proposing here. Could you explain the goal at a higher level?

was that related to my comment "How about resolve_domains_as_base_type for clarity?"?
if so, I was only referring to the name of the comment directive. The name of the SQL CTE is okay as-is

> Change the types to include Domain in [category type](https://github.com/andrew-w-ross/pg_graphql/blob/feature/domain-types/sql/load_sql_context.sql#L105-L112) and adding it to the [TypeCategory](https://github.com/andrew-w-ross/pg_graphql/blob/feature/domain-types/src/sql_types.rs#L86C1-L93C2) enum. Also create a Domain struct and add it to [TypeDetails](https://github.com/andrew-w-ross/pg_graphql/blob/feature/domain-types/src/sql_types.rs#L111C1-L116C2).

yep

> Revert the [replace functions](https://github.com/andrew-w-ross/pg_graphql/blob/feature/domain-types/src/sql_types.rs#L546-L617) in load_sql_context. Mapping the domain type in the [type details function](https://github.com/andrew-w-ross/pg_graphql/blob/feature/domain-types/src/sql_types.rs#L629-L647).

yep

> Finally adding another match block to for Domain in the [to_graphql_type](https://github.com/andrew-w-ross/pg_graphql/blob/feature/domain-types/src/graphql.rs#L1624-L1639) function.

yep

> As for the tests, you're happy with them as is so if they pass we're good right?

they looked good on my first pass. I'll re-review when you're ready but if any need to be adjusted I'm happy to do it

File: sql/load_sql_context.sql (hunk 8 above)
Comment (olirice) [reply sql/load_sql_context.sql:28]: yeah a formatter would be great but I haven't found one that I'm happy with. they tend to mangle unusual stuff like hierarchies of `jsonb_build_object`

throw the contents of that file into https://sqlformat.darold.net to see for yourself

File: test/expected/domain_field_types.out (hunk 10 above)
Comment (olirice) [reply test/expected/domain_field_types.out:1091]: just for documenting behavior, but if its too generic I might make a note to improve in the future

File: sql/load_sql_context.sql (hunk 8 above)
Comment (andrew-w-ross) [reply sql/load_sql_context.sql:28]: Point taken, that was unreadable.

File: src/sql_types.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (andrew-w-ross) [reply src/sql_types.rs:618]: > was that related to my comment "How about resolve_domains_as_base_type for clarity?"?

Nope, just changing the config key.

> sorry I didn't understand what you're proposing here. Could you explain the goal at a higher level?

Type category tends to map to a type off of the context query.  Enum to [enums](https://github.com/supabase/pg_graphql/blob/12b7aefae990c7a16157917e20386954cf5c63aa/sql/load_sql_context.sql#L55),  Composite to [composites](https://github.com/supabase/pg_graphql/blob/12b7aefae990c7a16157917e20386954cf5c63aa/sql/load_sql_context.sql#L130C7-L130C7) and Table to [tables](https://github.com/supabase/pg_graphql/blob/12b7aefae990c7a16157917e20386954cf5c63aa/sql/load_sql_context.sql#L237C12-L237C12). I was just going to do the same for domain types.

**Comment by olirice** [comment]
@andrew-w-ross I've converted this to a draft to keep the PRs clear. Please mark it "ready for review" once your ready

**Comment by olirice** [comment]
closing this for now. please re-open when you're ready
